pkg/eventing/sources/v1alpha1/apiserver_client.go

```diff
@@ -156,6 +159,43 @@ func (b *APIServerSourceBuilder) Resources(resources []v1alpha1.ApiServerResourc
 	return b
 }
 
+// AddResource which should be streamed
+func (b *APIServerSourceBuilder) AddResource(version string, kind string, isController bool) *APIServerSourceBuilder {
+	resources := b.apiServerSource.Spec.Resources
+	if resources == nil {
+		resources = []v1alpha1.ApiServerResource{}
+		b.apiServerSource.Spec.Resources = resources
+	}
+	resourceRef := v1alpha1.ApiServerResource{
+		APIVersion: version,
+		Kind:       kind,
+		Controller: isController,
+	}
+	b.apiServerSource.Spec.Resources = append(resources, resourceRef)
+	return b
+}
+
+// RemoveResource which should be streamed
+func (b *APIServerSourceBuilder) RemoveResource(version string, kind string, isController bool) (*APIServerSourceBuilder, error) {
+	resources := b.apiServerSource.Spec.Resources
+	if resources == nil {
+		resources = []v1alpha1.ApiServerResource{}
+		b.apiServerSource.Spec.Resources = resources
+	}
+	resourceRef := v1alpha1.ApiServerResource{
+		APIVersion: version,
+		Kind:       kind,
+		Controller: isController,
+	}
+	for i, k := range resources {
+		if reflect.DeepEqual(k, resourceRef) {
+			resources = append(resources[:i], resources[i+1:]...)
+			return b, nil
+		}
+	}
+	return b, fmt.Errorf("cannot find resource %s:%s:%t to remove", version, kind, isController)
```

Comment:
If the resource could be found, the function will return at the line #193 without error.
If the resource could not be found after traversing the whole structure, the function will return at line #196 with an error.
Could you help to explain more of your comment ?

Reply:
@maximilien  is right, chaining on works on methods which return a single return value, so returning two values makes no sense in this pattern.

But as we discussed in the other PR having mutating methods on a builder is an anti-pattern (as then the builder turns more into a regular domain objects). Maybe we can solve this use case differently, too ?

If not, we could agree that removing a non-existing resource is just a no-op and gets ignored, so that you can return directly the builder again.